pkgs/applications/editors/emacs/build-support/trivial.nix

```diff
@@ -2,9 +2,19 @@
 
 { callPackage, lib, ... }@envargs:
 
-args:
+let
+  libBuildHelper = import ./lib-build-helper.nix { inherit lib; };
+in
 
-callPackage ./generic.nix envargs ({
+libBuildHelper.extendMkDerivation { } (callPackage ./generic.nix envargs) (finalAttrs:
+
+{ pname
+, version
+, src
+, ...
+}@args:
+
+{
   buildPhase = ''
```

Comment:
I decide to ignore the user provided "previous" `args.buildPhase` because I consider `buildPhase` as the key component of this build helper.

That being said, users can still change `buildPhase` using `overrideAttrs` like this:

```nix
(trivialBuild {
  ...
}).overrideAttrs (finalAttrs: prevAttrs: {
  buildPhase = ...
})
```

The `buildPhase` and `installPhase` of other two build helpers, `elpaBuild` and `melpaBuild`, also follow this design.

 @ShamrockLee I see your related [comment][1] about factor lang.  Any thoughts here?

[1]: https://github.com/NixOS/nixpkgs/pull/287852#pullrequestreview-2208970569

Reply:
I'm unfamiliar with emacs packages, so commenting on the indispensability of the "key component" phases might be difficult.

A (counter) example for `buildGoModule` is that `apptainer` and `singularity` use the `make` and `make install` instead of `go build` and Golang-specific installation.

https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/virtualization/singularity/generic.nix#L275-L285

Reply:
> The user provided "previous" args.buildPhase is likely a mistake.

I am not sure if we should outsmart the user here.
You argue that there is an alternative overriding route for those really interested in shooting their own feet, but arguably this is more complicated.

Eval-time warnings are nicer, given that they can be turned off for those that swear to know what they are doing.

Reply:
Update:

*Nixpkgs Reference Manual* states that

> However, typically one only wants to add some commands to a phase, e.g. by defining postInstall or preFixup, as skipping some of the default actions may have unexpected consequences.

https://nixos.org/manual/nixpkgs/stable/#sec-stdenv-phases

IIUC, the statement means that one should typically use the pre- and post-attributes instead of overriding the whole phase. So, I would consider overriding the whole phase as a "use at your own risk" feature, even without additional warnings.

Reply:
> > Instead of ignoring `args.buildPhase`, maybe it is better to emit a warning at eval time
> 
> It is worth mentioning that such warnings are not suppressible (one can catch `throw`- based errors with `tryEval`, but not the warnings) and will appear when people `nix search` anything.

Presumably no elisp packages in Nixpkgs need their own `buildPhase` so such warnings are only seen by a small amount users with third party packages.

Note that a new argument can be easily introduced to control that warning.

> I am not sure if we should outsmart the user here.
You argue that there is an alternative overriding route for those really interested in shooting their own feet, but arguably this is more complicated.

👍

>  I would consider overriding the whole phase as a "use at your own risk" feature, even without additional warnings.

👍 I feel the same now.